internal/sqlsmith: TestGenerateParse failed

internal/sqlsmith.TestGenerateParse [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4019016&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4019016&tab=artifacts#/) on master @ [3471dc50db8406f3faa638940492dbe0cab53435](https://github.com/cockroachdb/cockroach/commits/3471dc50db8406f3faa638940492dbe0cab53435):


```
	defaultdb.public.seed AS tab_91
SELECT
	4:::INT8 AS col_192,
	1:::INT8 AS col_193,
	9:::INT8 AS col_194,
	8.0:::FLOAT8 AS col_195,
	2.0:::FLOAT8 AS col_196,
	(localtimestamp(1:::INT8::INT8):::DATE::DATE::DATE - 5:::INT8::INT8)::DATE AS col_197,
	NULL AS col_198,
	'2000-01-01 03:00:00+00:00':::TIMESTAMPTZ AS col_199,
	'00:00:02':::INTERVAL AS col_200,
	true AS col_201,
	6:::DECIMAL AS col_202,
	'D':::STRING AS col_203,
	NULL AS col_204,
	'00000000-0000-0000-0000-000000000009':::UUID AS col_205,
	'::6/0':::INET AS col_206,
	'9':::JSONB AS col_207,
	NULL AS col_208;

STMT: 61
COMMIT TRANSACTION;

STMT: 62
WITH
	with_20 (col_209)
		AS (
			SELECT
				*
			FROM
				(VALUES (7.0:::FLOAT8), (COALESCE(NULL, 9.0:::FLOAT8)), (NULL), (0.0:::FLOAT8), (5.0:::FLOAT8))
					AS tab_92 (col_209)
		),
	with_21 (col_210) AS (SELECT * FROM (VALUES (2:::INT8), (NULL)) AS tab_93 (col_210))
SELECT
	tab_94._string AS col_211,
	tab_94._date AS col_212,
	tab_95._bool AS col_213,
	tab_94.crdb_internal_mvcc_timestamp AS col_214,
	tab_94._jsonb AS col_215,
	tab_94._bytes AS col_216
FROM
	defaultdb.public.seed@[0] AS tab_94 LEFT JOIN defaultdb.public.seed@[0] AS tab_95 ON false
WHERE
	tab_94._bool
ORDER BY
	tab_94._string ASC, tab_95._timestamptz, tab_94._float4 DESC;

    sqlsmith_test.go:188: ALTER DATABASE defaultdb: at or near "EOF": syntax error
--- FAIL: TestGenerateParse (4.34s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestGenerateParse.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

internal/sqlsmith.TestGenerateParse [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4018014&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4018014&tab=artifacts#/) on master @ [3471dc50db8406f3faa638940492dbe0cab53435](https://github.com/cockroachdb/cockroach/commits/3471dc50db8406f3faa638940492dbe0cab53435):


```
COMMIT TRANSACTION;

STMT: 92
SELECT
	tab_153.col_279 AS col_281, tab_154.col_280 AS col_282
FROM
	(VALUES (false), (NULL), (true), (false), (NULL), (e'JA\x0fK@"':::STRING NOT LIKE e'/P+B\x16\x15':::STRING))
		AS tab_153 (col_279),
	(VALUES (e'\\x74954649ada21ed4':::BYTES)) AS tab_154 (col_280)
WHERE
	tab_153.col_279
ORDER BY
	tab_154.col_280, tab_154.col_280, tab_153.col_279 DESC;

STMT: 93
WITH
	with_44 (col_285)
		AS (
			SELECT
				NULL AS col_285
			FROM
				(VALUES ('"':::STRING, false), ('I':::STRING, false), (NULL, true), ('9':::STRING, false))
					AS tab_155 (col_283, col_284)
		)
SELECT
	tab_170.col_302 AS col_304, tab_171.col_303 AS col_305
FROM
	(VALUES ('E':::STRING:::NAME), (e'R^\x1a-\fN':::STRING:::NAME), (e'"\x1eb\x7f%**_':::STRING:::NAME))
		AS tab_170 (col_302)
	LEFT JOIN (
			VALUES
				(
					ARRAY[
						22203:::INT8,
						(-19332):::INT8,
						3621:::INT8,
						20024:::INT8,
						(-17587):::INT8,
						4604:::INT8,
						31683:::INT8
					]
				)
		)
			AS tab_171 (col_303) ON true;

    sqlsmith_test.go:188: ALTER DATABASE defaultdb: at or near "EOF": syntax error
W211231 12:13:04.914206 10932 kv/kvserver/intentresolver/intent_resolver.go:793  [-] 538  failed to gc transaction record: could not GC completed transaction anchored at /Table/SystemConfigSpan/Start: node unavailable; try another peer
W211231 12:13:04.914297 10300 jobs/registry.go:858  [n1] 539  canceling all adopted jobs due to stopper quiescing
W211231 12:13:04.914704 10170 sql/sqlliveness/slinstance/slinstance.go:224  [n1] 540  exiting heartbeat loop
--- FAIL: TestGenerateParse (0.56s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestGenerateParse.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

internal/sqlsmith.TestGenerateParse [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4018018&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4018018&tab=artifacts#/) on master @ [3471dc50db8406f3faa638940492dbe0cab53435](https://github.com/cockroachdb/cockroach/commits/3471dc50db8406f3faa638940492dbe0cab53435):


```
						(e'\\x47':::BYTES, e'\x1fx\x04t\t+':::STRING)
				)
					AS tab_15 (col_32, col_33)
		),
	with_8 (col_34) AS (SELECT * FROM (VALUES (3059000227:::OID), (570607391:::OID)) AS tab_16 (col_34))
SELECT
	tab_18.col_37 AS col_38
FROM
	(
		VALUES
			(
				(
					SELECT
						3464636140:::OID AS col_36
					FROM
						(
							VALUES
								(
									'01070000E0E61000000500000001060000E0E61000000200000001030000C00100000008000000A05CBE5F52F118C026BFA68A998D42C0BC0CC53CD1B4F64120D0450CFC72F3C19824207927E93040B39F65E6EC6644C0AC85FF4AC1C4E341A69D9D126335F9C1F04CDDDFCFB057400F63BEE53B2055C0684E7DBE297CD141459F8F54596B01C2A8D4567336AF5F4011291BDC393342C02449FA3BAD42FB41F25CC74D2D4BF9C14898BDB4FDBE5940CC36CA3974A830C0F432FAEFDF87E1411697A7E816B6FA4132AB88163ACF6140D804A16F62F92D40B40ABE118328F94134560DA48B1902C2E4F4380F7F386240224C6C3AEB9F54403CEA70D5D022D7C1C4D187B364A6EB41A05CBE5F52F118C026BFA68A998D42C0BC0CC53CD1B4F64120D0450CFC72F3C101030000C0010000000900000031360672671760C0D2C2CF36825650C0463EDB15F3C5F84142D9E3586B16FCC1704642F95133424052BCE947AF1256C0D321A5162110F0C12F674F30FF1202C2D835DD6144A04B40978AC7E1C2AC4CC074F68478C09500C2A4894CA7AB3002C23484EE2F82A05840FF766CE0832B53C06CC8465FD4FFD6C1ECD44F91660AE6410857E06698855A403A2B6DB40C0251C01838D835ABF50042BCD5F3F5C505F7C104AC98A5DE4565406CB462357F2633C0801FD7B4F9090042A84E31A742E0E541546197DE3F5F5140A06975FC15ED5540B0B8B127BA8EE7418960BF3EF845F6C112A363B0026D66C080DFD6DA97A25240BA22DCA6C1DBECC1F01A2B66B2A0DD4131360672671760C0D2C2CF36825650C0463EDB15F3C5F84142D9E3586B16FCC101010000E0E610000018FAB37D53DE33C02806B5C817BD2D4057779A8C3D4F00C22238DC81E347F24101010000E0E6100000C8E8A304759B5340443F3831CDDA5040DAAADFA42649F6417C90FF7F2B91ED4101040000E0E61000000700000001010000C06A82A6F9D4976140A42B1D95D2625440B673DB266F4202C218A01CCFDE51D44101010000C0A68A4BDEEAF563C09470E624CEAD4940B46D8C92903CF641FAD5CF2F4B2602C201010000C0FEEF756EEC996040BCF18CA6EF6643C05CE004120A810242786063C971FEE54101010000C0348DBFD8FE8E57C08E40959C6B5F52C03D31A09DACD0F0C1002CA84571F66E4101010000C0187083BC058F38408CC3E6B38D0F47C06481D39F53A9E1419E52488B6B13F8C101010000C0BE67AA89485960C0CA109FF3649845C050D912590701EBC1A01FC5DDE459D14101010000C0C85C936459AC3D40F20369E11FDF50C07021C6651ACED8411685AFA60891E2C101040000E0E61000000700000001010000C0223EA68D01AD6040007EAF03421ED63F5CFC1F0B1763E841FC447440C885ED4101010000C0F834A256B7973B40907FD3D809022840D80AE759A22DE8C118FA99C9E321D4C101010000C050BE8C1F098356C0D0C3ED42829A1C4060C81D9F5FA6F34110F7A057EF18E94101010000C054E6DC91884B5B40E0AE9D2CD39A414058248451FB50DA411A6072314818004201010000C08B43513318E756C030171A550E3C4B405C63866DBA19E941BA3497AAF171F4C101010000C0903F283B674D4440B86091C0346253C0B82C9ADB0039FEC1AF0102D5A8D4F6C101010000C0E8C4CB296EE23E40046BB989367A4340B83603930DB9E641C8AD6C3AC3E7F741':::GEOGRAPHY
								)
						)
							AS tab_17 (col_35)
					LIMIT
						1:::INT8
				)
			),
			(0:::OID),
			(1799084014:::OID)
	)
		AS tab_18 (col_37)
ORDER BY
	tab_18.col_37 ASC
LIMIT
	68:::INT8;

STMT: 19
ROLLBACK TRANSACTION;

STMT: 20
SELECT true AS col_39;

STMT: 21
SELECT 1073468986:::OID AS col_40;

    sqlsmith_test.go:188: ALTER DATABASE defaultdb: at or near "EOF": syntax error
W211231 12:15:48.785791 11050 jobs/registry.go:858  [n1] 581  canceling all adopted jobs due to stopper quiescing
I211231 12:15:48.786861 11830 2@rpc/context.go:742  [n1] 582  closing [{n1: 127.0.0.1:38515 (default)} {n0: 127.0.0.1:38515 (default)}]
W211231 12:15:48.787911 10887 sql/sqlliveness/slinstance/slinstance.go:224  [n1] 583  exiting heartbeat loop
I211231 12:15:48.844138 10720 kv/kvclient/kvcoord/transport_race.go:107  [-] 584  transport race promotion: ran 27 iterations on up to 270 requests
--- FAIL: TestGenerateParse (1.89s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestGenerateParse.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

internal/sqlsmith.TestGenerateParse [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4024059&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4024059&tab=artifacts#/) on master @ [fa9480a8b661b31a3c5e4c3428595b3e7c320700](https://github.com/cockroachdb/cockroach/commits/fa9480a8b661b31a3c5e4c3428595b3e7c320700):


```
					),
				with_19 (col_108)
					AS (
						SELECT
							*
						FROM
							(
								VALUES
									(NULL),
									(e'\\x5bfa':::BYTES),
									(e'\\x051aa28aa4d054':::BYTES),
									(e'\\x3c422f':::BYTES),
									(e'\\xf15a2d688d04641a30':::BYTES)
							)
								AS tab_69 (col_108)
						UNION ALL
							SELECT
								*
							FROM
								(
									VALUES
										(e'\\x0afe05f9458b345924':::BYTES),
										(e'\\x58':::BYTES),
										(e'\\x':::BYTES),
										(e'\\x':::BYTES),
										(e'\\x055aa359':::BYTES)
								)
									AS tab_70 (col_109)
					)
			SELECT
				cte_ref_5.col_106 AS col_110
			FROM
				with_18 AS cte_ref_5
			ORDER BY
				cte_ref_5.col_106 DESC, cte_ref_5.col_106, cte_ref_5.col_106 DESC, cte_ref_5.col_106 DESC
		)
		AND st_equals(
				'010700000000000000':::GEOMETRY::GEOMETRY,
				'01040000000300000001010000002202AA9B9BF90042968274A681CCFDC1010100000080B06D2A17C39E417CC49A70D91DE7410101000000B8D3963F02A2FD412DF49D39AD3AF0C1':::GEOMETRY::GEOMETRY
			)::BOOL
	);

STMT: 59
COMMIT TRANSACTION;

STMT: 60
ROLLBACK TRANSACTION;

    sqlsmith_test.go:188: ALTER DATABASE defaultdb: at or near "EOF": syntax error
--- FAIL: TestGenerateParse (2.09s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestGenerateParse.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/internal/sqlsmith/alter.go
+++ b/pkg/internal/sqlsmith/alter.go
@@ -462,14 +462,14 @@ func makeAlterDatabaseDropRegion(s *Smither) (tree.Statement, bool) {
 }
 
 func makeAlterSurvivalGoal(s *Smither) (tree.Statement, bool) {
-	const numSurvivalGoals = 2
-	// Only {1, 2} are valid values for SurvivalGoal.
-	// 1. SurvivalGoalRegionFailure
-	// 2. SurvivalGoalZoneFailure
-	//
-	// 0. SurvivalGoalDefault is not valid as a value to alter to.
-	randInt := rand.Intn(numSurvivalGoals) + 1
-	survivalGoal := tree.SurvivalGoal(randInt)
+	// Only SurvivalGoalRegionFailure and SurvivalGoalZoneFailure are valid
+	// values for SurvivalGoal. SurvivalGoalDefault is not valid in an
+	// AlterDatabaseSurvivalGoal AST node.
+	survivalGoals := [...]tree.SurvivalGoal{
+		tree.SurvivalGoalRegionFailure,
+		tree.SurvivalGoalZoneFailure,
+	}
+	survivalGoal := survivalGoals[rand.Intn(len(survivalGoals))]
 
 	ast := &tree.AlterDatabaseSurvivalGoal{
 		Name:         tree.Name("defaultdb"),
@@ -479,13 +479,18 @@ func makeAlterSurvivalGoal(s *Smither) (tree.Statement, bool) {
 }
 
 func makeAlterDatabasePlacement(s *Smither) (tree.Statement, bool) {
-	const DataPlacementTypes = 3
-	randInt := rand.Intn(DataPlacementTypes)
-	dataPlacementType := tree.DataPlacement(randInt)
+	// Only DataPlacementDefault and DataPlacementRestricted are valid values
+	// for Placement. DataPlacementUnspecified is not valid in an
+	// AlterDatabasePlacement AST node.
+	dataPlacements := [...]tree.DataPlacement{
+		tree.DataPlacementDefault,
+		tree.DataPlacementRestricted,
+	}
+	dataPlacement := dataPlacements[rand.Intn(len(dataPlacements))]
 
 	ast := &tree.AlterDatabasePlacement{
 		Name:      tree.Name("defaultdb"),
-		Placement: dataPlacementType,
+		Placement: dataPlacement,
 	}
 
 	return ast, true

